README.rst
===============
marketanalysis
===============

A fast, efficient Python library for generating market holidays and trading days based on countries'
security exchanges. It aims to make determining whether a specific date is a market 
holiday as fast and flexible as possible.


Example Usage
-------------

.. code-block:: python

    from datetime import date
    from marketanalysis import marketholidays
    from marketanalysis import markettradingdays

    us_marketholidays = marketholidays.UnitedStates()
    # or:
    # us_marketholidays = marketholidays.US()
    # or:
    # us_marketholidays = marketholidays.CountryHoliday('US')
    # or, for specific prov / states:
    # us_marketholidays = marketholidays.CountryHoliday('US', prov=None, state='CA')

    date(2015, 1, 1) in us_marketholidays  # True
    date(2015, 1, 2) in us_marketholidays  # False

    # The Holiday class will also recognize strings of any format
    # and int/float representing a Unix timestamp
    '2014-01-01' in us_marketholidays  # True
    '1/1/2014' in us_marketholidays    # True
    1388597445 in us_marketholidays    # True

    us_marketholidays.get('2014-01-01')  # "New Year's Day"

    us_marketholidays['2014-01-01': '2014-01-03']  # [date(2014, 1, 1)]

    us_pr_marketholidays = marketholidays.UnitedStates()  # or marketholidays.US(...), or marketholidays.CountryHoliday('US', state='PR')

    # some marketholidays are only present in parts of a country
    '2018-01-06' in us_marketholidays     # False
    '2018-01-06' in us_pr_marketholidays  # True

    # Easily create custom Holiday objects with your own dates instead
    # of using the pre-defined countries/states/provinces available
    custom_marketholidays = marketholidays.MarketHolidayBase()
    # Append custom holiday dates by passing:
    # 1) a dict with date/name key/value pairs,
    custom_marketholidays.append({"2015-01-01": "New Year's Day"})
    # 2) a list of dates (in any format: date, datetime, string, integer),
    custom_marketholidays.append(['2015-07-01', '07/04/2015'])
    # 3) a single date item
    custom_marketholidays.append(date(2015, 12, 25))

    date(2015, 1, 1) in custom_marketholidays  # True
    date(2015, 1, 2) in custom_marketholidays  # False
    '12/25/2015' in custom_marketholidays      # True

    # For more complex logic like 4th Monday of January, you can inherit the
    # HolidayBase class and define your own _populate(year) method. See below
    # documentation for examples.


Install
-------

The latest stable version can always be installed or updated via pip:

.. code-block:: bash

    $ pip install marketholidays==0.11
    
Clone repo with git bash:

 .. code-block:: bash 
 
   $ git clone https://github.com/OmoMicheal/marketanalysis.git


Available Countries
-------------------

=================== =========== =============================================================
Country             ISO code    Securities/Exchanges Available
=================== =========== =============================================================
Canada              CA/CAN      Canadian Securities Exchange
Nigeria             NG/NGA      Nigerian Stock Exchange, NSE
SouthAfrica         ZA/ZAF      Johannesburg Stock Exchange
UnitedStates        US/USA      New York Stock Exchange (NYSE);
                                National Association of Securities 
                                Dealers Automated Quotations (NASDAQ)
=================== =========== =============================================================


API
---

class marketholidays.MarketHolidayBase(years=[], expand=True, observed=True, prov=None, state=None)
    The base class used to create holiday country classes.

Parameters:

years
    An iterable list of integers specifying the years that the Holiday object
    should pre-generate. This would generally only be used if setting *expand*
    to False. (Default: [])

expand
    A boolean value which specifies whether or not to append marketholidays in new
    years to the marketholidays object. (Default: True)

observed
    A boolean value which when set to True will include the observed day of a
    holiday that falls on a weekend, when appropriate. (Default: True)

prov
    A string specifying a province that has unique statutory marketholidays.
    (Default: Canada='ON')

state
    A string specifying a state that has unique statutory marketholidays.
    (Default: UnitedStates=None)

Methods:

get(key, default=None)
    Returns a string containing the name of the holiday(s) in date ``key``, which
    can be of date, datetime, string, unicode, bytes, integer or float type. If
    multiple marketholidays fall on the same date the names will be separated by
    commas

get(key, default=None)
    Returns a string containing the name of the holiday(s) in date ``key``, which
    can be of date, datetime, string, unicode, bytes, integer or float type. If
    multiple marketholidays fall on the same date the names will be separated by
    commas

get_list(key)
    Same as ``get`` except returns a ``list`` of holiday names instead of a comma
    separated string

get_named(name)
    Returns a ``list`` of marketholidays matching (even partially) the provided name
    (case insensitive check)

pop(key, default=None)
    Same as ``get`` except the key is removed from the holiday object

pop_named(name)
    Same as ``pop`` but takes the name of the holiday (or part of it) rather than
    the date

update/append
    Accepts dictionary of {date: name} pairs, a list of dates, or even singular
    date/string/timestamp objects and adds them to the list of marketholidays


-------------

.. code-block:: python

    # Simplest example possible

    >>> from datetime import date
    >>> from marketanalysis import marketholidays
    >>> date(2021, 4, 2) in marketholidays.US()
    True
    >> date(2021, 1, 2) in marketholidays.US()
    False

    # However, this is not efficient because it is initializing a new market holiday 
    # object and generating a list of all the holidays in 2021 during each comparison.

    # It is more efficient to create the object only once

    >>> us_marketholidays = marketholidays.US()
    >>> date(2021, 4, 2) in us_marketholidays
    True
    >> date(2021, 1, 2) in us_marketholidays
    False

    # You can pick whichever you prefer of the following two.

    >>> marketholidays.UnitedStates() == marketholidays.US()
    True


    # Let's print out the market holidays in 2021 specific to NYSE and NASDAQ

    >>> for date, name in sorted(marketholidays.US(state='CA', years=2021).items()):
    >>>     print(date, name)
    2021-01-01 New Year's Day
    2021-01-18 Martin Luther King, Jr. Day
    2021-02-15 Presidents' Day
    2021-04-02 Good Friday
    2021-05-31 Memorial Day
    2021-07-04 Independence Day
    2021-07-05 Independence Day (Observed)
    2021-09-06 Labour Day
    2021-10-11 Thanksgiving
    2021-12-24 Christmas Day (Observed)
    2021-12-25 Christmas Day
    2021-12-31 New Year's Day (Observed)
    
    
    from marketholidays import NG
    >>> for date, name in sorted(NG(years=2021).items()):
    >>>    print(date, name)
    
    2021-01-01 New Year's Day
    2021-04-02 Good Friday
    2021-04-05 Easter Monday
    2021-05-01 Workers’ Day
    2021-05-03 Workers’ Day (Observed)
    2021-06-12 Democracy Day
    2021-06-14 Democracy Day (Observed)
    2021-10-01 Independence Day
    2021-12-25 Christmas Day
    2021-12-26 Boxing Day
    2021-12-27 Christmas Day (Observed)
    2021-12-31 New Year's Day (Observed)



    from marketholidays import CA
    >>> for date, name in sorted(CA(years=2022).items()):
    >>>    print(date, name)
    
    2021-12-31 New Year's Day (Observed)
    2022-01-01 New Year's Day
    2022-02-21 Family Day
    2022-04-15 Good Friday
    2022-05-23 Victoria Day
    2022-07-01 Canada Day
    2022-08-01 Civic Holiday
    2022-09-05 Labour Day
    2022-10-10 Thanksgiving
    2022-12-25 Christmas Day
    2022-12-26 Boxing Day
    2022-12-27 Christmas Day (Observed)



CHANGES
Version 0.1
===========

Released April 12, 2021

- Initial release
- Support for Python >= 3
- Includes market holidays and trading days for Canada, South Africa, Nigeria, and United States.


marketanalysis/marketholidays/countries/south_africa.py
# -*- coding: utf-8 -*-

#  marketanalysis
#  ----------------
#  A fast, efficient Python library for generating country, province and state
#  specific sets of marketholidays on the fly. It aims to make determining whether a
#  specific date is a holiday as fast and flexible as possible.
#
#  Author:  MichealOmojola <momojola@aust.edu.ng>
#  Website: https://github.com/OmoMicheal/trading_days
#  License: MIT (see LICENSE file)
#  Version: 0.1 (April 7, 2021)


from datetime import date
from dateutil.relativedelta import relativedelta as rd
from dateutil.relativedelta import MO, TU, FR
from dateutil.easter import easter


from marketholidays.marketholiday_base import MarketHolidayBase

# from datetime import date, datetime

# from dateutil.easter import easter
# from dateutil.relativedelta import relativedelta as rd

# from holidays.constants import FRI, SUN
# from holidays.constants import JAN, MAR, APR, MAY, JUN, JUL, AUG, SEP, OCT, DEC
# from holidays.holiday_base import HolidayBase


class SouthAfrica(MarketHolidayBase):
    def __init__(self, **kwargs):
        # https://www.tradinghours.com/markets/jse/holidays (Johannesburg Stock Exchange)
        self.country = "ZA"
        MarketHolidayBase.__init__(self, **kwargs)

    def _populate(self, year):
        # New Year's Day
        if year > 1909:
            name = "New Year's Day"
            self[date(year, 1, 1)] = name
            if self.observed and date(year, 1, 1).weekday() == 6:
                self[date(year, 1, 1) + rd(days=+1)] = name + " (Observed)"
            elif self.observed and date(year, 1, 1).weekday() == 5:
                # Add Dec 31st from the previous year without triggering
                # the entire year to be added
                expand = self.expand
                self.expand = False
                self[date(year, 1, 1) + rd(days=-1)] = name + " (Observed)"
                self.expand = expand
            # The next year's observed New Year's Day can be in this year
            # when it falls on a Friday (Jan 1st is a Saturday)
            if self.observed and date(year, 12, 31).weekday() == 4:
                self[date(year, 12, 31)] = name + " (Observed)"
        
        if year > 1994:
            name = "Human Rights Day"
            self[date(year, 3, 21)] = name
            if self.observed and date(year, 3, 21).weekday() == 5:
                self[date(year, 3, 21) + rd(weekday = MO(+1))] = name + " (Observed)"
            elif self.observed and date(year, 3, 21).weekday() == 6:
                self[date(year, 3, 21) + rd(weekday = MO(+1))] = name + " (Observed)"
        
        # Good Friday
        self[easter(year) + rd(weekday=FR(-1))] = "Good Friday"
        
        # Family Day
        if year >= 1995:
            self[easter(year) + rd(weekday=MO(+1))] = "Family Day"
        
        if year > 1994:
            name = "Freedom Day"
            self[date(year, 4, 27)] = name
            if self.observed and date(year, 4, 27).weekday() == 5:
                self[date(year, 4, 27) + rd(weekday = MO(+1))] = name + " (Observed)"
            elif self.observed and date(year, 4, 27).weekday() == 6:
                self[date(year, 4, 27) + rd(weekday = MO(+1))] = name + " (Observed)"

        if year > 1994:
            name = "Youth Day"
            self[date(year, 6, 16)] = name
            if self.observed and date(year,  6, 16).weekday() == 5:
                self[date(year,  6, 16) + rd(weekday = MO(+1))] = name + " (Observed)"
            elif self.observed and date(year,  6, 16).weekday() == 6:
                self[date(year,  6, 16) + rd(weekday = MO(+1))] = name + " (Observed)"

        if year > 1994:
            name = "National Women's Day"
            mm = 8; dd = 9
            self[date(year, mm, dd)] = name
            if self.observed and date(year, mm, dd).weekday() == 5:
                self[date(year, mm, dd) + rd(weekday = MO(+1))] = name + " (Observed)"
            elif self.observed and date(year, mm, dd).weekday() == 6:
                self[date(year, mm, dd) + rd(weekday = MO(+1))] = name + " (Observed)"

        if year > 1994:
            name = "Heritage Day"
            mm = 9; dd = 24
            self[date(year, mm, dd)] = name
            if self.observed and date(year, mm, dd).weekday() == 5:
                self[date(year, mm, dd) + rd(weekday = MO(+1))] = name + " (Observed)"
            elif self.observed and date(year, mm, dd).weekday() == 6:
                self[date(year, mm, dd) + rd(weekday = MO(+1))] = name + " (Observed)"

        if year > 1994:
            name = "Day of Reconciliation"
            mm = 12; dd = 16
            self[date(year, mm, dd)] = name
            if self.observed and date(year, mm, dd).weekday() == 5:
                self[date(year, mm, dd) + rd(weekday = MO(+1))] = name + " (Observed)"
            elif self.observed and date(year, mm, dd).weekday() == 6:
                self[date(year, mm, dd) + rd(weekday = MO(+1))] = name + " (Observed)"

        # Christmas Day
        if year >= 1910:
            name = "Christmas Day"
            mm = 12; dd = 25
            self[date(year, mm, dd)] = name
            if self.observed and date(year, mm, dd).weekday() == 5:
                self[date(year, mm, dd) + rd(weekday = MO(+1))] = name + " (Observed)"
            elif self.observed and date(year, mm, dd).weekday() == 6:
                self[date(year, mm, dd) + rd(weekday = TU(+1))] = name + " (Observed)"

        # Boxing Day
        if year >= 1910:
            name = "Boxing Day"
            mm = 12; dd = 26
            self[date(year, mm, dd)] = name
            if self.observed and date(year, mm, dd).weekday() == 5:
                self[date(year, mm, dd) + rd(weekday = MO(+1))] = name + " (Observed)"
            elif self.observed and date(year, mm, dd).weekday() == 6:
                self[date(year, mm, dd) + rd(weekday = TU(+1))] = name + " (Observed)"


class ZA(SouthAfrica):
    pass


class ZAF(SouthAfrica):
    pass
